### ISSUE TITLE
expects_dialect(sicstus) breaks op declarations in standard modules

### ISSUE THREAD

**Issue by dgelessus** [SWI-Prolog/swipl-devel#731]
Tested with SWI-Prolog built from source (44b4a32986cf35a525f413a38aa3ff658c7c1a48). This issue can be seen with `library(clpfd)` for example:

```prolog
?- expects_dialect(sicstus).
true.

?- use_module(library(clpfd)).
ERROR: <builddir>/home/library/clp/clpfd.pl:950:17: Syntax error: Operator expected
ERROR: <builddir>/home/library/clp/clpfd.pl:953:17: Syntax error: Operator expected
ERROR: <builddir>/home/library/clp/clpfd.pl:954:17: Syntax error: Operator expected
ERROR: <builddir>/home/library/clp/clpfd.pl:955:17: Syntax error: Operator expected
ERROR: <builddir>/home/library/clp/clpfd.pl:956:17: Syntax error: Operator expected
<... many more similar errors ...>
ERROR: <builddir>/home/library/clp/clpfd.pl:5443:14: Syntax error: Operator expected
ERROR: <builddir>/home/library/clp/clpfd.pl:5842:22: Syntax error: Operator expected
ERROR: <builddir>/home/library/clp/clpfd.pl:7359:
ERROR:    clpfd:parse_clpfd_clauses/1: Unknown procedure: clpfd:parse_clpfd/3
ERROR: <builddir>/home/library/clp/clpfd.pl:7360:
ERROR:    clpfd:parse_reified_clauses/1: Unknown procedure: clpfd:parse_reified/4
ERROR: <builddir>/home/library/clp/clpfd.pl:7361:
ERROR:    call_term_expansion/5: Unknown procedure: clpfd:make_matches/1
true.

?- X #= Y + 2, Y #= 5.
Warning: <builddir>/home/library/prolog_stack.pl:46: 
Warning:   library(lists) does not export nth1/3
ERROR: Unknown procedure: clpfd:clpfd_equal_/2
ERROR:   However, there are definitions for:
ERROR:         clpfd:clpfd_equal/2
ERROR: 
ERROR: In:
ERROR:   [12] clpfd:clpfd_equal_(_14264,_14270+2)
ERROR:   [11] clpfd:clpfd_equal(_14302,_14308+2) at <builddir>/home/library/clp/clpfd.pl:2795
ERROR:   [10] '<meta-call>'(user:user: ...) <foreign>
ERROR:    [9] toplevel_call(user:user: ...) at <builddir>/home/boot/toplevel.pl:1113
   Exception: (12) clpfd:clpfd_equal_(_6382, _6404+2) ? leap
```

As a workaround I can manually load the modules in question before `expects_dialect(sicstus)`:

```prolog
?- use_module(library(clpfd)).
true.

?- expects_dialect(sicstus).
true.

?- X #= Y + 2, Y #= 5.
X = 7,
Y = 5.
```

I think this is because of this goal expansion in `library(dialect/sicstus)`, which redirects all unqualified `op` declarations into the `user` module, to simulate the SICStus behavior of all operator declarations being global:

https://github.com/SWI-Prolog/swipl-devel/blob/44b4a32986cf35a525f413a38aa3ff658c7c1a48/library/dialect/sicstus.pl#L116-L128

However, when this goal expansion is applied to SWI standard modules, it makes them unable to see their own operator declarations, because according to [the `op/3` docs](https://www.swi-prolog.org/pldoc/doc_for?object=op/3):

> Files that are loaded from the SWI-Prolog directories resolve operators and predicates from this `system` module rather than `user`, [...]

A workaround would be to explicitly qualify all operator names in SWI standard modules (e. g. `:- op(700, xfx, clpfd:(cis)).`), which causes the goal expansion to ignore them, but this probably isn't a good solution. It would be better if the `op` goal expansion didn't affect SWI standard modules. Or perhaps `expects_dialect(sicstus)` should ignore SWI standard modules entirely?

**Comment by JanWielemaker** [comment]
Using `user:goal_expansion(...)` should fix that.  That used to have other problems, but these are resolved long ago.  Maybe it is also a good idea to make `prolog_load_context(dialect, X)` `swi` for all library modules.   Still, you get an error as the file search path is updated to load the SICStus libraries.

The typical usage is to use the expects_dialect/1 _after_ loading the libraries that you need.

**Comment by dgelessus** [comment]
> The typical usage is to use the expects_dialect/1 _after_ loading the libraries that you need.

That would mean putting an `expects_dialect` directive into every source file, right? (Well, every source file that needs SICStus emulation, which is basically every source file in this codebase 🙂) At the moment I simply put `:- expects_dialect(sicstus).` near the start of the main entry point module, before all of its `use_module` directives. This way all other modules loaded from there are also loaded with SICStus emulation, which is quite convenient.

We also have quite a few places in the code where `use_module` directives are used in the middle of the file and not all grouped in a single section at the top. For example, in a module with 50 predicates where only one of them uses `term_hash`, there might be a `:- use_module(library(terms), [term_hash/2]).` right before the one predicate declaration that uses `term_hash/2`. There are also a few more complex cases with `use_module` in `:- if.`/`:- endif.`, which aren't always trivial to move to the top of the file. Because of this, some SWI standard modules would still end up being loaded with SICStus emulation active, even if I would always put `:- expects_dialect(sicstus).` after the main group of `use_module`s.

**Comment by dgelessus** [comment]
> Using `user:goal_expansion(...)` should fix that.

Did some testing locally, and this seems to work relatively well - at least I'm no longer getting a ton of syntax errors in clpfd.pl. But with this change I'm running into another issue with code like this:

```prolog
:- use_module(library(clpfd),[fd_dom/2, fd_degree/2, fd_size/2]).
print_var(V) :- fd_dom(V,D), D \= inf..sup, fd_size(V,Sz), fd_degree(V,Dg),!,
   format('INT VAR: ~w:~w [sz:~w,dg:~w] ',[V,D,Sz,Dg]).
```

On SWI this causes a syntax error on the `inf..sup` part. The `..` operator is declared in `library(clpfd)`, but it's not being imported here, because `library(clpfd)` is loaded with an explicit list of imports that doesn't include `op(450, xfx, ..)`. On SICStus this isn't an issue, because operators aren't module-sensitive, so they don't need to be listed in `use_module/2` import lists.

Another issue: on SICStus, once an operator has been declared, it can be used by *any* code, even code that doesn't (directly or indirectly) import the module declaring the operator. For example, if a module `main` imports modules `one` and `two`, and module `one` imports `library(clpfd)`, then module `two` on SICStus is able to use the `..` operator, even though `two` hasn't imported `library(clpfd)`. Thankfully I haven't yet found any code that relies on this questionable behavior, but I wouldn't be surprised if there is some that does...

The first issue (but not the second) could probably be solved by adding another goal expansion that rewrites all `use_module/2` calls to always import all operators exported by the imported module. Both issues could be solved by changing the `op` goal expansion so that all operators are declared both in the current module and `user` - e. g. `:- op(450, xfx, ..).` would become `:- op(450, xfx, ..), op(450, xfx, user:(..)).`.

**Comment by JanWielemaker** [comment]
> That would mean putting an `expects_dialect` directive into every source file, right?

Yes.  It might be possible to solve this.   Let m think a bit.

> Both issues could be solved by changing the `op` goal expansion so that all operators are declared both in the current module and `user` - e. g. `:- op(450, xfx, ..).` would become `:- op(450, xfx, ..), op(450, xfx, user:(..)).`.

Interesting thought.   You could try that.   Then you still needs the expansion is `system`.   That can, at least theoretically, break SWI-Prolog's modules.   You seem to understand most of the mechanisms :smile: 

This is rather complicated :cry:

**Comment by dgelessus** [comment]
Some more testing later...

I wasn't able to get the second solution to work (redeclaring all operators both in the module that declares them and in `user`). The problem is that in SWI code, exported operators usually aren't declared using standalone `:- op(Prio,Assoc,Name).` directives, but using `op(Prio,Assoc,Name)` terms in the exports list of the `module/2` directive. I wasn't able to rewrite those operator declarations - it seems that `system:goal_expansion/2` and `system:term_expansion/2` don't get called for the `module/2` directives in SWI standard library files. They do get called for `module/2` directives in user code, but that doesn't help here.

The first solution seems to work quite well though (rewriting `use_module/2` directives to always import all exported operators). I found in the [`load_files/2`](https://www.swi-prolog.org/pldoc/man?predicate=load_files/2) docs that putting `op(_,_,_)` in an import list will import all operators exported by the module. This is very helpful here - with this feature, the rewrite can be implemented in just 3 lines:

```prolog
user:goal_expansion(use_module(Module,Imports),
		    use_module(Module,[op(_,_,_)|Imports])) :-
	\+ memberchk(op(_,_,_),Imports).
```

Of course this doesn't solve the problem of SICStus code that expects operators from a module that it doesn't import. But I haven't found any code in our codebase that requires this behavior, and IMHO it's not too much to ask that when code uses operators from a module, it should import that module...

I'll send a PR with the operator fixes I've implemented so far - they seem to work well with our code. Thank you for all the help so far!
